export_realm: Add RealmExport model.

prakhar1144 opened the pull request:
This PR adds a `RealmExport` model.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/worker/deferred_work.py:
```diff
@@ -126,60 +125,72 @@ def failure_processor(event: dict[str, Any]) -> None:
+                    ).decode()
+                    export_event.save(update_fields=["extra_data"])
+
+            if export_row.date_started is not None:
```
I think you want a migration to populate the new table from `RealmAuditLog`; you might still need a bit of code here to look up the object properly, but it'd mean all other code paths can just read from `RealmExport` (like how does one have clean `actions` code listing all the exports if one didn't do a migration?).

prakhar1144 replied:
Yeah, this was still a WIP. A lot of other bugs were still there :)

Userpanya commented:
T083243Japan169074
Warranty Certificate

prakhar1144 commented on zerver/migrations/0588_realmexport.py:
```diff
@@ -0,0 +1,118 @@
+            realm=audit_log.realm,
+            acting_user=audit_log.acting_user,
+            # TODO: No way to determine if it was a EXPORT_FULL_WITHOUT_CONSENT ?
+            type=EXPORT_PUBLIC if audit_log.acting_user else EXPORT_FULL_WITH_CONSENT,
```
I think we don't have the data to find if the export made via command line was `EXPORT_FULL_WITH_CONSENT` or `EXPORT_FULL_WITHOUT_CONSENT`

timabbott replied:
Sounds possible. Probably this is the best we can do for now.

prakhar1144 commented on zerver/views/realm_export.py:
```diff
@@ -97,20 +95,17 @@ def get_realm_exports(request: HttpRequest, user: UserProfile) -> HttpResponse:
-            id=export_id, realm=user.realm, event_type=AuditLogEventType.REALM_EXPORTED
-        )
-    except RealmAuditLog.DoesNotExist:
+        export_row = RealmExport.objects.get(id=export_id)
```
Is it possible that a client makes an API request to delete `export_id` (which refers to id of `RealmAuditLog`) and at the same instant migration completes and the request now tries to delete an object with incorrect `id` (of `RealmExport`) ?

Do we discard all the API requests for the period of migration & django restart during deployment ?

Understanding the exact order in which things happen during deployment would be helpful here, I think!

timabbott replied:
It's very unlikely that a relevant API request of this form happens during the migration, just given their frequency. But generally migrations are applied before the server restarts, rather than the other way around.

prakhar1144 commented:
This PR is ready for review.

* Wrote the migration code to backfill RealmExport table.

timabbott commented on zerver/migrations/0588_realmexport.py:
```diff
@@ -0,0 +1,118 @@
+            export_path=export_path,
+        )
+
+    RealmAuditLog.objects.filter(event_type=REALM_EXPORTED).delete()
```
Started https://chat.zulip.org/#narrow/stream/3-backend/topic/RealmExport.20model.20migrations/near/1952551 to discuss whether it's appropriate to delete these.

prakhar1144 replied:
Updated code to NOT delete historical `RealmAuditLog` entries.

timabbott commented on zerver/views/realm_export.py:
```diff
@@ -68,8 +65,11 @@ def export_realm(request: HttpRequest, user: UserProfile) -> HttpResponse:
+        type=RealmExport.EXPORT_PUBLIC,
+        acting_user=user,
+        date_requested=timezone_now(),
     )
```
Reading this, I kinda feel like we should have a `PositiveSmallIntegerField` enum state variable, initially `REQUESTED`, so the state machine isn't entirely tracked via these timestamps.

prakhar1144 replied:
Updated with:
```
    REQUESTED = 1
    STARTED = 2
    SUCCEEDED = 3
    FAILED = 4
    DELETED = 5
    status = models.PositiveSmallIntegerField(default=REQUESTED)
```

* Removed the `date_started` field.
* We still need the `date_requested` field for limit check. (5 exports within a week)

---
Question to further simplify the code:

Do you think it's important to differentiate between `STARTED` and `REQUESTED` state ? I think we can simply mark it as `STARTED` when the user makes a request for export (and not wait for concerned worker to actually start the export process). It'll be only helpful if for some reason we want to track those export requests which the worker hasn't yet picked up.

prakhar1144 replied:
Question to further simplify the code:

Do you think it's important to differentiate between `STARTED` and `REQUESTED` state ? I think we can simply mark it as `STARTED` when the user makes a request for export (and not wait for concerned worker to actually start the export process). It'll be only helpful if for some reason we want to track those export requests which the worker hasn't yet picked up.

timabbott replied:
Yes, it is helpful to distinguish for dealing with bugs, in that something in the `STARTED` state we expect to be currently running, while in the `REQUESTED` state we expect it to be in the queue.

prakhar1144 commented on zerver/lib/export.py:
```diff
@@ -2441,35 +2455,79 @@ def get_consented_user_ids(realm: Realm) -> set[int]:
-    return public_url
+        print("Calculating SHA-256 checksum of tarball...")
+
+        # TODO: We can directly use 'hashlib.file_digest'. (New in Python 3.11)
```
https://docs.python.org/3/library/hashlib.html#hashlib.file_digest

prakhar1144 commented on zerver/lib/export.py:
```diff
@@ -2441,35 +2455,79 @@ def get_consented_user_ids(realm: Realm) -> set[int]:
+                size = f.readinto(buf)
+                if size == 0:
+                    break
+                sha256_hash.update(view[:size])
```
https://github.com/python/cpython/blob/656b7a3c83c79f99beac950b59c47575562ea729/Lib/hashlib.py#L230

prakhar1144 commented:
Updates in the recent push:
* Updated backfill code in migration file [as per discussion](https://github.com/zulip/zulip/pull/31760#discussion_r1783327831).
* Added `status` field as suggested -- https://github.com/zulip/zulip/pull/31760#discussion_r1783344357
* Added `sha256summ_hex`, `tarball_size_bytes` fields as [discussed in CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/RealmExport.20model.20migrations/near/1952613)

Questions:
* https://github.com/zulip/zulip/pull/31760#discussion_r1786083750
*

timabbott commented on zerver/migrations/0595_add_realmexport_table_and_backfill.py:
```diff
@@ -0,0 +1,97 @@
+            date_started=timestamp_to_datetime(started_timestamp),
+            date_failed=timestamp_to_datetime(failed_timestamp),
+            date_deleted=timestamp_to_datetime(deleted_timestamp),
+            export_path=export_path,
```
This doesn't set an initial value of `status`. I think probably we can compute it from the `date_` fields...

timabbott commented:
This lgtm aside from the note above and potentially storing `date_started/date_completed` timestamps in the `RealmExport` model for future analysis.

prakhar1144 commented:
Posted a few [questions in CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/RealmExport.20model.20migrations/near/1955172)

timabbott commented on zerver/views/realm_export.py:
```diff
@@ -97,20 +96,17 @@ def get_realm_exports(request: HttpRequest, user: UserProfile) -> HttpResponse:
         raise JsonableError(_("Export still in progress"))
-    do_delete_realm_export(audit_log_entry)
+    do_delete_realm_export(export_row)
     return json_success(request)
```
We probably will want to add a new `RealmAuditLog` type for deleting an export, such that `acting_user` can be tracked properly. This can be a quick follow-up.

timabbott commented:
Merged, thanks @prakhar1144!
